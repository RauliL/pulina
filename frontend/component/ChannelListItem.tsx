import classNames from 'classnames';
import * as React from 'react';

export interface Props {
  name: string;
  isActive: boolean;
  hasUnreadMessages: boolean;
  onSelect: () => void;
}

const getLinkClassName = (props: Props) => classNames(
  'p-2',
  'd-block',
  'text-secondary',
  'text-decoration-none',
  {
    'bg-white': props.isActive,
    'text-body': props.isActive,
  },
);

const getChannelNameClassName = (props: Props) => classNames({
  'font-weight-bold': props.hasUnreadMessages,
  'text-body': props.isActive,
  'text-white': !props.isActive && props.hasUnreadMessages,
  'text-secondary': !props.isActive && !props.hasUnreadMessages,
});

const ChannelListItem: React.SFC<Props> = (props) => (
  <li>
    <a
      href="#"
      className={getLinkClassName(props)}
      onClick={(event) => {
        event.preventDefault();
        props.onSelect();
      }}
    >
      #
      &nbsp;
      <span className={getChannelNameClassName(props)}>
        {props.name.substring(1)}
      </span>
    </a>
  </li>
);

export default ChannelListItem;
