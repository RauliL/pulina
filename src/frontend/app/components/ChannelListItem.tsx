import classNames from "classnames";
import React, { FunctionComponent } from "react";

export type ChannelListItemProps = {
  hasUnreadHighlights: boolean;
  hasUnreadMessages: boolean;
  isActive: boolean;
  name: string;
  onSelect: () => void;
};

const getLinkClassName = (props: ChannelListItemProps) =>
  classNames("p-2", "d-block", "text-secondary", "text-decoration-none", {
    "bg-white": props.isActive,
    "text-body": props.isActive,
  });

const getChannelNameClassName = (props: ChannelListItemProps) =>
  classNames({
    "font-weight-bold": props.hasUnreadMessages || props.hasUnreadHighlights,
    "text-body": props.isActive,
    "text-white":
      !props.isActive && props.hasUnreadMessages && !props.hasUnreadHighlights,
    "text-warning": !props.isActive && props.hasUnreadHighlights,
    "text-secondary": !props.isActive && !props.hasUnreadMessages,
  });

const ChannelListItem: FunctionComponent<ChannelListItemProps> = (props) => (
  <li data-testid="ChannelListItem">
    <a
      href="#"
      className={getLinkClassName(props)}
      onClick={(event) => {
        event.preventDefault();
        props.onSelect();
      }}
    >
      # &nbsp;
      <span className={getChannelNameClassName(props)}>
        {props.name.substring(1)}
      </span>
    </a>
  </li>
);

ChannelListItem.displayName = "ChannelListItem";

export default ChannelListItem;
