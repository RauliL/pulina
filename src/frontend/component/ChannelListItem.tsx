import classNames from "classnames";
import React, { FunctionComponent } from "react";

export type Props = {
  name: string;
  isActive: boolean;
  hasUnreadMessages: boolean;
  hasUnreadHighlight: boolean;
  onSelect: () => void;
};

const getLinkClassName = (props: Props) =>
  classNames("p-2", "d-block", "text-secondary", "text-decoration-none", {
    "bg-white": props.isActive,
    "text-body": props.isActive,
  });

const getChannelNameClassName = (props: Props) =>
  classNames({
    "font-weight-bold": props.hasUnreadMessages || props.hasUnreadHighlight,
    "text-body": props.isActive,
    "text-white":
      !props.isActive && props.hasUnreadMessages && !props.hasUnreadHighlight,
    "text-warning": !props.isActive && props.hasUnreadHighlight,
    "text-secondary": !props.isActive && !props.hasUnreadMessages,
  });

const ChannelListItem: FunctionComponent<Props> = (props) => (
  <li>
    <a
      href="#"
      className={getLinkClassName(props)}
      onClick={(event) => {
        event.preventDefault();
        props.onSelect();
      }}
    >
      # &nbsp;
      <span className={getChannelNameClassName(props)}>
        {props.name.substring(1)}
      </span>
    </a>
  </li>
);

export default ChannelListItem;
