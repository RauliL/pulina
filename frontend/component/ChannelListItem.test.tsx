import { mount, shallow } from 'enzyme';
import { defaultTo, noop } from 'lodash';
import * as React from 'react';

import ChannelListItem, { Props } from './ChannelListItem';

describe('<ChannelListItem/>', () => {
  const renderComponent = (props: Partial<Props> = {}) => (
    <ChannelListItem
      name={defaultTo(props.name, '#test-channel')}
      isActive={defaultTo(props.isActive, false)}
      hasUnreadMessages={defaultTo(props.hasUnreadMessages, false)}
      hasUnreadHighlight={defaultTo(props.hasUnreadHighlight, false)}
      onSelect={defaultTo(props.onSelect, noop)}
    />
  );
  const mountComponent = (props: Partial<Props> = {}) => mount(
    renderComponent(props),
  );

  it('should render without throwing exception', () => {
    expect(shallow(renderComponent())).toMatchSnapshot();
  });

  it('should render <li/>', () => {
    expect(mountComponent().find('li')).toHaveLength(1);
  });

  it('should render <a/>', () => {
    expect(mountComponent().find('a')).toHaveLength(1);
  });

  it('should render channel name without hash', () => {
    const component = mountComponent();

    expect(component.find('a').find('span').text()).toBe('test-channel');
  });

  it('should invoke callback when channel is being selected', () => {
    const onSelect = jest.fn();
    const component = mountComponent({ onSelect });

    component.find('a').simulate('click');
    expect(onSelect).toBeCalled();
  });
});
