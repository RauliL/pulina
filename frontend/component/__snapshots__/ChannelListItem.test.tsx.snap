// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelListItem/> should render without throwing exception 1`] = `
<li>
  <a
    className="p-2 d-block text-secondary text-decoration-none"
    href="#"
    onClick={[Function]}
  >
    # Â 
    <span
      className="text-secondary"
    >
      test-channel
    </span>
  </a>
</li>
`;
